apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java

```diff
@@ -0,0 +1,249 @@
+                            headerValue = rewriteBuffer.toString();
+                        }
+                        sampleRate = rounded;
+                    }
+                } catch (NumberFormatException e) {
+                    // silently ignored
+                }
+            }
+        }
+
+        tracestate.add(headerValue);
+    }
+
+    /**
+     * Sets value for trace state. Provided rate and string value are assumed to be correct and consistent
+     *
+     * @param rate       sample rate
+     * @param rateString rate written as a string, used to minimize allocation
+     * @throws IllegalStateException    if sample rate has already been set
+     * @throws IllegalArgumentException if rate has an invalid value
+     */
+    public void set(double rate, String rateString) {
+        if (!Double.isNaN(sampleRate)) {
+            // sample rate is set either explicitly from this method (for root transactions)
+            // or through upstream header, thus there is no need to change after. This allows to only
+            // write/rewrite headers once
+            throw new IllegalStateException("sample rate has already been set from headers");
+        }
+        sampleRate = rate;
+        header.setLength(0);
+        header.append(FULL_PREFIX);
+        header.append(rateString);
+        tracestate.add(header);
+    }
+
+    /**
+     * @return sample rate between 0.0 and 1.0, or {@link Double#NaN} if not set
+     */
+    public double getSampleRate() {
+        return sampleRate;
+    }
+
+    @Nullable
+    public String toTextHeader() {
+        if (tracestate.isEmpty()) {
+            return null;
+        } else {
+            return TextTracestateAppender.INSTANCE.join(tracestate, sizeLimit);
+        }
+    }
+
+    @Override
+    public void resetState() {
+        sampleRate = Double.NaN;
+        sizeLimit = DEFAULT_SIZE_LIMIT;
+        rewriteBuffer.setLength(0);
+        tracestate.clear();
+    }
+
+    public void setSizeLimit(int limit) {
+        if(!tracestate.isEmpty()) {
+            throw new IllegalStateException("can't change size limit once headers have been added");
+        }
+        this.sizeLimit = limit;
+    }
+
+    /**
+     * Internal appender uses a per-thread StringBuilder instance to concatenate the tracestate header.
+     * This allows ot limit actual memory usage to be linear to the number of active threads which
+     * is assumed to be far less than the number of active in-flight transactions.
+     */
+    private static class TextTracestateAppender {
+
+        private static final TextTracestateAppender INSTANCE = new TextTracestateAppender();
+        private final ThreadLocal<StringBuilder> tracestateBuffer = new ThreadLocal<StringBuilder>();
+
+        private TextTracestateAppender() {
+        }
+
+        @Nullable
+        public String join(List<? extends CharSequence> tracestate, int tracestateSizeLimit) {
```

Comment:
[minor] The optimization we lost in `TraceContext` (just forwarding a header as it was received):
```suggestion
        public String join(List<? extends CharSequence> tracestate, int tracestateSizeLimit) {
          if (tracestate.size() == 1 && tracestate.get(0) instanceof String) {
            return (String) tracestate.get(0);
          }
```

Reply:
Thanks for catching this, there was one corner case to apply here though when tracestate is transmitted as-is and larger that the size limit we have to truncate it :-)